src/common/Throttle.cc

```diff
@@ -95,26 +103,31 @@ bool Throttle::_wait(int64_t c)
   utime_t start;
   bool waited = false;
   if (_should_wait(c) || !cond.empty()) { // always wait behind other waiters.
-    Cond *cv = new Cond;
-    cond.push_back(cv);
-    waited = true;
-    ldout(cct, 2) << "_wait waiting..." << dendl;
-    if (logger)
-      start = ceph_clock_now();
-
-    do {
-      cv->Wait(lock);
-    } while (_should_wait(c) || cv != cond.front());
-
-    ldout(cct, 2) << "_wait finished waiting" << dendl;
-    if (logger) {
-      utime_t dur = ceph_clock_now() - start;
-      logger->tinc(l_throttle_wait, dur);
+    {
+      auto cv = cond.insert(cond.end(), new Cond);
+      auto w = make_scope_guard([this, cv]() {
+	  delete *cv;
+	  cond.erase(cv);
+	});
+      waited = true;
+      ldout(cct, 2) << "_wait waiting..." << dendl;
+      if (logger)
+	start = ceph_clock_now();
+
+      do {
+	(*cv)->Wait(lock);
+      } while ((_should_wait(c) || cv != cond.begin()) && !shutting_down);
+      if (shutting_down) {
+	shutdown_clear.SignalOne();
+	throw waiting_destructed("Throttle destructed while we waited.");
```

Comment:
Mostly we don't like the way it messes around with control flow. We also use the google style guide as our baseline and nobody has ever made a persuasive argument for breaking with it in this case.

(Also, it's polite to users of our libraries not to make them play with exceptions.)